src/Uno.Foundation/Collections/NullableKeyDictionary.cs

```diff
@@ -0,0 +1,193 @@
+﻿#nullable enable
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Uno.Foundation.Collections
+{
+	/// <summary>
+	/// Represents a variant of dictionary that allows adding a value for a null key.
+	/// </summary>
+	/// <typeparam name="TKey">Key.</typeparam>
+	/// <typeparam name="TValue">Value.</typeparam>
+	internal class NullableKeyDictionary<TKey, TValue> : IDictionary<TKey, TValue>
+		  where TKey : class
+	{
+		private TValue _nullValue = default!;
+		private bool _containsNullValue;
+		private readonly Dictionary<TKey, TValue> _dictionary;
+
+		public NullableKeyDictionary() =>
+			_dictionary = new Dictionary<TKey, TValue>();
+
+		public NullableKeyDictionary(IEqualityComparer<TKey> comparer) =>
+			_dictionary = new Dictionary<TKey, TValue>(comparer);
+
+		public bool ContainsKey(TKey? key) => key == null ? _containsNullValue : _dictionary.ContainsKey(key);
+
+		public void Add(TKey? key, TValue value)
+		{
+			if (key == null)
+			{
+				_nullValue = value;
+				_containsNullValue = true;
+			}
+			else
+			{
+				_dictionary[key] = value;
+			}
+		}
+
```

Comment:
This implementation isn't consistent with IDictionary documentation:

https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.idictionary-2.add?view=net-5.0

`Add` method is supposed to throw if the element already exists.